### PR TITLE
Allow update remove labels

### PR DESCRIPTION
**Pull request by DvirDukhan**
This PR allows calling the "REMOVE" keyword on node labels

```
MATCH (n) REMOVE n:L RETURN labels(n)
MATCH (n) REMOVE n:L1:L2 RETURN labels(n)
```

- [x] code
- [x] TCK
- [x] FLOW

### CODE REVIEW

File: src/ast/ast_build_reference_map.c [hunk 1]
```diff
@@ -220,6 +220,34 @@ static void _AST_MapSetClauseReferences(AST *ast, const cypher_astnode_t *set_cl
...
+	ASSERT(cypher_astnode_type(remove_item) == CYPHER_AST_REMOVE_LABELS);
+	const cypher_astnode_t *identifier = cypher_ast_remove_labels_get_identifier(remove_item);
+	ASSERT(cypher_astnode_type(identifier) == CYPHER_AST_IDENTIFIER);
+	const char *alias = cypher_ast_identifier_get_name(identifier);
+	_AST_UpdateRefMap(ast, alias);
+}
+
+static void _AST_MapRemoveItemReferences(AST *ast, const cypher_astnode_t *remove_item) {
+	const cypher_astnode_type_t type = cypher_astnode_type(remove_item);
+	if(type == CYPHER_AST_REMOVE_LABELS) {
```
Comment (AviAvni) [inline src/ast/ast_build_reference_map.c:234]: remove the if and just have
ASSERT(type == CYPHER_AST_REMOVE_LABELS)

File: src/graph/graph.c [hunk 2]
```diff
@@ -616,6 +616,39 @@ static void _Graph_LabelNode
...
+		// remove matrix at position [id, id]
+		RG_Matrix m = Graph_GetLabelMatrix(g, l);
+		info = RG_Matrix_removeElement_BOOL(m, id, id);
+		ASSERT(info == GrB_SUCCESS);
+
+		// remove this label from this node's set of labels
+		info = RG_Matrix_removeElement_BOOL(nl, id, l);
+		ASSERT(info == GrB_SUCCESS);
+
+		// a node with 'label' has just been created, update statistics
```
Comment (AviAvni) [inline src/graph/graph.c:647]: update comment

File: src/graph/graph.h [hunk 3]
```diff
@@ -133,6 +133,14 @@ void Graph_LabelNode
 	uint label_count
 );
 
+void Graph_RemoveLabelNode
```
Comment (AviAvni) [inline src/graph/graph.h:136]: add documentation

File: src/graph/graph_hub.c [hunk 4]
```diff
@@ -279,23 +279,20 @@ int UpdateEntityProperties
...
 }
 
-int UpdateNodeLabels
+
+static int _SetNodeLabels
 (
 	GraphContext *gc,            // graph context to update the entity
-	Node *node,                  // the node to be updated
-	rax *labels                  // labels to update
+	Node *node,            		 // the node to be updated
```
Comment (AviAvni) [inline src/graph/graph_hub.c:286]: align spaces

File: src/graph/graph_hub.c [hunk 5]
```diff
@@ -279,23 +279,20 @@ int UpdateEntityProperties
...
+	rax *add_labels     	     // labels to add
 ) {
-	ASSERT(gc != NULL);
-	ASSERT(node != NULL);
```
Comment (AviAvni) [inline src/graph/graph_hub.c:289]: why this was removed?

File: src/graph/graph_hub.c [hunk 6]
```diff
@@ -326,8 +323,70 @@ int UpdateNodeLabels
 
 	// Update label matrixes.
 	Graph_LabelNode(gc->g, node->id ,label_ids, label_count);
+	
+	// cleanup
+	array_free(label_ids);
```
Comment (AviAvni) [inline src/graph/graph_hub.c:328]: is this a leak we didn't noticed?

File: src/graph/graph_hub.c [hunk 7]
```diff
@@ -326,8 +323,70 @@ int UpdateNodeLabels
...
+	Graph_RemoveLabelNode(gc->g, node->id ,label_ids, array_len(label_ids));
+	// cleanup
 	array_free(label_ids);
+
+}
+
+int UpdateNodeLabels
+(
+	GraphContext *gc,            // graph context to update the entity
+	Node *node,            		 // the node to be updated
```
Comment (AviAvni) [inline src/graph/graph_hub.c:382]: align space

File: src/graph/graph_hub.h [hunk 8]
```diff
@@ -84,5 +84,6 @@ int UpdateNodeLabels
 (
 	GraphContext *gc,            // graph context to update the entity
 	Node *node,            		 // the node to be updated
```
Comment (AviAvni) [inline src/graph/graph_hub.h:86]: align space

**Review by AviAvni — COMMENTED** [review]
see comments

File: src/graph/graph_hub.c (hunk 5 above)
Comment (DvirDukhan) [reply src/graph/graph_hub.c:289]: I think it was just moved...

File: src/graph/graph_hub.c (hunk 6 above)
Comment (DvirDukhan) [reply src/graph/graph_hub.c:328]: this is GitHub showing new changes from some reason

File: src/ast/ast_shared.h [hunk 9]
```diff
@@ -66,7 +66,8 @@ typedef struct {
 // Context describing an update expression.
 typedef struct {
 	PropertySetCtx *properties; // properties to set
-	rax *labels;                // Labels to update the entity with (add/remove according to the update mode).
+	rax *add_labels;            // Labels to add to the entity.
+	rax *remove_labels;         // Labels to remove from the entity.
```
Comment (AviAvni) [inline src/ast/ast_shared.h:70]: merge to one max and the value indicate if we remove or add the label

File: src/graph/graph.h [hunk 10]
```diff
@@ -133,6 +133,15 @@ void Graph_LabelNode
 	uint label_count
 );
 
+// Remove assocoation of node with labels by removing the value 1 from label matrix L at position [id,id]
```
Comment (AviAvni) [inline src/graph/graph.h:136]: add the remove of the entry in node label matrix NLM[node_id, label_id]
update also in line 127

File: tests/flow/test_entity_update.py [hunk 11]
```diff
@@ -385,3 +385,49 @@ def test_27_set_label_on_merge(self):
...
+        self.validate_node_labels(graph, labels, 0)
+
+    def test_29_mix_add_and_remove_node_labels(self):
+        graph.delete()
+        graph.query("CREATE (:Foo)")
+        labels_to_add = ["Bar"]
+        labels_to_remove = ["Foo"]
+        self.validate_node_labels(graph, labels_to_remove, 1)
+
+        result = graph.query(f"MATCH (n:Foo) SET n:{':'.join(labels_to_add)} REMOVE n:{':'.join(labels_to_remove)} RETURN n")
```
Comment (AviAvni) [inline tests/flow/test_entity_update.py:411]: permutations
SET n:L REMOVE n:L
REMOVE n:L SET n:L

File: tests/flow/test_entity_update.py [hunk 12]
```diff
@@ -385,3 +385,49 @@ def test_27_set_label_on_merge(self):
...
+        graph.delete()
+        labels_to_remove = ["Foo"]
+        self.validate_node_labels(graph, labels_to_remove, 0)
+
+        result = graph.query(f"MERGE (n:{':'.join(labels_to_remove)})  REMOVE n:{':'.join(labels_to_remove)} RETURN n")
+        self.env.assertEqual(result.labels_added, 1)
+        self.validate_node_labels(graph, labels_to_remove, 0)
+
+    def test_30_syntax_error_remove_labels_on_match_on_create(self):
+        queries = ["MERGE (n) ON MATCH REMOVE n:Foo RETURN n", "MERGE (n) ON CREATE REMOVE n:Foo RETURN n"]
```
Comment (AviAvni) [inline tests/flow/test_entity_update.py:426]: MATCH ()-[n]->() REMOVE n:R

**Review by AviAvni — CHANGES_REQUESTED** [review]

**Review by AviAvni — APPROVED** [review]

**Pull request merged**